### PR TITLE
Add support for creating, getting, and resolving predictions

### LINKED ISSUE #178: [2.X Feature Request] Support for the Predictions API
In early May, Twitch released the Polls and Predictions API as an initial beta then released it fully towards the end of the month

Having this functionality in the library would be extremely useful for streamers and/or moderators wanting to automate creating and ending predictions for their channels.

https://dev.twitch.tv/docs/api/reference#create-prediction

### PR DESCRIPTION
**Pull request by raechelmaelstrom**
Resolves #178

### CODE REVIEW

**Comment by raechelmaelstrom** [comment]
This doesn't have all the fields of the prediction data structure, but just wanting to check myself before I go any further.  Works with my test client to create, get and resolve predictions.

File: twitchio/models.py [hunk 1]
```diff
@@ -581,3 +582,36 @@ def __init__(self, http: "TwitchHTTP", data: dict):
...
+
+    def __init__(self, http: "TwitchHTTP", data: dict):
+        self.user = PartialUser(http, data["broadcaster_id"], data["broadcaster_name"])
+        self.prediction_id: str = data["id"]
+        self.title: str = data["title"]
+        self.winning_outcome_id: str = data["winning_outcome_id"]
+        self.outcomes: List[PredictionOutcome] = [PredictionOutcome(http, x) for x in data["outcomes"]]
+        self.prediction_window: int = data["prediction_window"]
+        self.prediction_status: str = data["status"]
+        self.created_at: str = data["created_at"]
```
Comment (IAmTomahawkx) [inline twitchio/models.py:599]: These should be converted to datetime objects.

File: twitchio/models.py [hunk 2]
```diff
@@ -581,3 +582,36 @@ def __init__(self, http: "TwitchHTTP", data: dict):
...
+
+class PredictionOutcome:
+
+    __slots__ = ("outcome_id", "title", "channel_points", "color")
+
+    def __init__(self, http: "TwitchHTTP", data: dict):
+        self.outcome_id: str = data["id"]
+        self.title: str = data["title"]
+        self.channel_points: int = data["channel_points"]
+        self.color: str = data["color"]
```
Comment (IAmTomahawkx) [inline twitchio/models.py:614]: Alias colour to color

File: twitchio/user.py [hunk 3]
```diff
@@ -609,6 +609,86 @@ async def fetch_videos(self, period="all", sort="time", type="all", language=Non
         data = await self._http.get_videos(user_id=str(self.id), period=period, sort=sort, type=type, language=language)
         return [Video(self._http, x, self) for x in data]
 
+    async def end_prediction(self, token: str, prediction_id: str, status: str, winning_outcome_id: str = None):
```
Comment (IAmTomahawkx) [inline twitchio/user.py:612]: Missing return type

File: twitchio/user.py [hunk 4]
```diff
@@ -609,6 +609,86 @@ async def fetch_videos(self, period="all", sort="time", type="all", language=Non
...
+
+        data = await self._http.patch_prediction(
+                token,
+                broadcaster_id=str(self.id),
+                prediction_id=prediction_id,
+                status=status,
+                winning_outcome_id=winning_outcome_id)
+        return Prediction(self._http, data[0])
+
+    async def get_predictions(self, token: str, prediction_id: str = None):
```
Comment (IAmTomahawkx) [inline twitchio/user.py:637]: Missing return type

File: twitchio/user.py [hunk 5]
```diff
@@ -609,6 +609,86 @@ async def fetch_videos(self, period="all", sort="time", type="all", language=Non
...
+        """
+        from .models import Prediction
+
+        data = await self._http.get_predictions(
+                token,
+                broadcaster_id=str(self.id),
+                prediction_id=prediction_id)
+        return [Prediction(self._http, d) for d in data]
+
+    async def create_prediction(self, token: str, title: str, blue_outcome: str, pink_outcome: str, prediction_window: int):
```
Comment (IAmTomahawkx) [inline twitchio/user.py:661]: Missing return type

File: twitchio/user.py [hunk 6]
```diff
@@ -609,6 +609,86 @@ async def fetch_videos(self, period="all", sort="time", type="all", language=Non
...
+        """|coro|
+        Creates a prediction for the channel.
+
+        Parameters
+        -----------
+        token: :class:`str`
+            An oauth token with the channel:manage:predictions scope
+        title: :class:`str`
+            Title for the prediction (max of 45 characters)
+        outcomes: :class:`List[str]`
```
Comment (IAmTomahawkx) [inline twitchio/user.py:671]: Docstring doesn't match signature

File: twitchio/http.py [hunk 7]
```diff
@@ -520,6 +520,64 @@ async def update_reward_redemption_status(
...
+            "outcomes": [
+                {
+                    "title": blue_outcome,
+                },
+                {
+                    "title": pink_outcome,
+                }
+            ],
+        }
+        logging.debug(f"Sending prediction {body}")
```
Comment (IAmTomahawkx) [inline twitchio/http.py:575]: (Doing) logging here isn't consistent with the rest of the library

**Review by IAmTomahawkx — CHANGES_REQUESTED** [review]

**Comment by IAmTomahawkx** [comment]
Thanks for the pr, just a couple things that need changing

**Comment by raechelmaelstrom** [comment]
Okay this one is ready for another look!  I had some trouble with the date time stuff because apparently twitch is giving me dates of the format: 2021-09-07T01:21:27.725465223Z which python did not like.  Seems like different parts of the API all have different time stamp formats though.

**Comment by IAmTomahawkx** [comment]
The api is very inconsistent. I'm pretty sure they try to be as inconsistent as possible tbh. I'll review this again in a few hours, thanks

**Review by IAmTomahawkx — APPROVED** [review]

**Pull request merged**